### PR TITLE
DCAC-47 - Controller & Validation unit tests for POST request

### PR DESCRIPTION
**Pull request by rmcmurray-ch**
Fulfils [DCAC-47](https://companieshouse.atlassian.net/browse/DCAC-47)

[DCAC-47]: https://companieshouse.atlassian.net/browse/DCAC-47?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by rmcmurray-ch** [comment]
PR for controller & validation tests only, work ongoing on integration tests

File: src/test/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        final ResponseEntity<Object> response = controller.createItemGroup(requestId, itemGroupData);
+
+        assertThat(response.getStatusCode(), is(HttpStatus.CONFLICT));
+
+        ItemGroupData returnedItemGroupData = (ItemGroupData) response.getBody();
+        assertNotNull(returnedItemGroupData);
+        assertThat(returnedItemGroupData, is(itemGroupData));
+    }
+    /**
+     * Fair weather with all validation triggered DTO
```
Comment (VaughanJackson123) [inline src/test/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupControllerTest.java:146]: I find this comment confusing. Are we saying it passes all validation?

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/validation/ItemGroupsValidator.java [hunk 2]
```diff
@@ -1,97 +1,142 @@
...
         }
     }
 
-    private void validateDescriptionIdentifier(ItemGroupData dto, List<String> errors) {
-        // TODO - DCAC-47 - for future validation.
+    private void validateItemDescriptionIdentifier(ItemGroupData dto, List<String> errors) {
+        for (Item item : dto.getItems()) {
+            String descriptionIdentifier = item.getDescriptionIdentifier();
+
+            if (isNull(ItemDescriptionIdentifier.getEnumValue(descriptionIdentifier))) {
```
Comment (VaughanJackson123) [inline src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/validation/ItemGroupsValidator.java:113]: Did you look into using an annotation like `@ValueOfEnum(enumClass = ItemDescriptionIdentifier.class)`?

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java [hunk 3]
```diff
@@ -13,96 +13,116 @@
...
-    private ResponseEntity<Object> buildCreateSuccessResponse(final ItemGroup savedItem) {
+        logger.getLogger().info(REQUEST_ID_PREFIX, dataMap.getLogMap());
+    }
+    /**
+     * @return ResponseEntity.status(CREATED)
+     */
+    private ResponseEntity<Object> buildCreateSuccessResponse(String requestId,
+                                                              final ItemGroup savedItem) {
         DataMap dataMap = new DataMap.Builder()
+            .xRequestId(requestId)
```
Comment (VaughanJackson123) [inline src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java:80]: Should we be using `xRequestId` or `requestId`?

**Review by VaughanJackson123 — APPROVED** [review]

**Comment by rmcmurray-ch** [comment]
@VaughanJackson123 thanks Vaughan, Martin is adding suggested fixes to next push 👍 much appreciate your help

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java (hunk 3 above)
Comment (VaughanJackson123) [reply src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java]: Sorry, I see my question may have been unclear here. I was really asking about whether we should be using the DataMap Builder `xRequestId()` or `requestId()` method there. They appear to be equivalent?

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java [hunk 4]
```diff
@@ -13,96 +15,119 @@
 import uk.gov.companieshouse.logging.util.DataMap;
 
 import java.util.List;
-import java.util.Map;
 
 import static org.springframework.http.HttpStatus.*;
 
 @RestController
+@TestConfiguration
```
Comment (VaughanJackson123) [inline src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java:22]: This annotation should not appear on an implementation class.

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java [hunk 5]
```diff
@@ -13,96 +15,119 @@
...
             final ItemGroup savedItem = itemGroupsService.createItemGroup(itemGroupData);
-            return buildCreateSuccessResponse(savedItem);
-
-        } catch (Exception ex) {
-            return buildErrorResponse(ex, itemGroupData);
+            return buildCreateSuccessResponse(xRequestId, savedItem);
+        }
+        catch (Exception ex) {
+            return buildErrorResponse(xRequestId, ex, itemGroupData);
         }
```
Comment (VaughanJackson123) [inline src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java:69]: This catch all handling risks misrepresenting an issue such as an NPE as a 400 Bad Request?

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/model/ItemCostProductType.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
+package uk.gov.companieshouse.itemgroupworkflowapi.model;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public enum ItemCostProductType {
+    CERTIFICATE("productTypeName"),
```
Comment (VaughanJackson123) [inline src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/model/ItemCostProductType.java:9]: `certificate`?

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java [hunk 7]
```diff
@@ -13,96 +15,119 @@
...
     }
 
-    private void logRequestId(String requestId) {
-        Map<String, Object> logMap = logger.createLogMap();
-        logMap.put(REQUEST_ID_LOG_KEY, requestId);
-        logger.getLogger().info("create item group: request id", logMap);
-    }
+    private void logRequestId(String xRequestId) {
+        DataMap dataMap = new DataMap.Builder()
+            .xRequestId(xRequestId)
```
Comment (VaughanJackson123) [inline src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java:74]: Following discussions, we have removed this `xRequestId()` method from the DataMap Builder as a duplicate key logging key for the request ID (just now).

**Comment by rmcmurray-ch** [comment]
Closing due to size of PR and DCAC-78's changes being included. New PR will be opened covering just DCAC-47.

**Pull request closed without merging**